// element-react/dist/npm/es5/src/upload/Cover.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _libs = require('../../libs');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Cover = function (_Component) {
  (0, _inherits3.default)(Cover, _Component);

  function Cover(props) {
    (0, _classCallCheck3.default)(this, Cover);

    var _this = (0, _possibleConstructorReturn3.default)(this, (Cover.__proto__ || Object.getPrototypeOf(Cover)).call(this, props));

    _this.state = {
      dragOver: false
    };
    return _this;
  }

  (0, _createClass3.default)(Cover, [{
    key: 'handleDragover',
    value: function handleDragover(e) {
      e.preventDefault();
      if (!this.props.disabled) {
        this.setState({ dragOver: true });
      }
    }
  }, {
    key: 'handleDragleave',
    value: function handleDragleave(e) {
      e.preventDefault();
      this.setState({ dragOver: false });
    }
  }, {
    key: 'onDrop',
    value: function onDrop(e) {
      if (this.props.disabled) return;
      e.preventDefault();
      this.setState({ dragOver: false });
      this.props.onFile(e.dataTransfer.files);
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var dragOver = this.state.dragOver;

      return _react2.default.createElement(
        'div',
        {
          className: this.classNames({
            'el-upload-dragger': true,
            'is-dragover': dragOver
          }),
          onDrop: function onDrop(e) {
            return _this2.onDrop(e);
          },
          onDragOver: function onDragOver(e) {
            return _this2.handleDragover(e);
          },
          onDragLeave: function onDragLeave(e) {
            return _this2.handleDragleave(e);
          }
        },
        this.props.children
      );
    }
  }]);
  return Cover;
}(_libs.Component);

Cover.defaultProps = {
  onFile: Function
};
var _default = Cover;
exports.default = _default;


Cover.propTypes = {
  onFile: _libs.PropTypes.func,
  disabled: _libs.PropTypes.bool
};
;

var _temp = function () {
  if (typeof __REACT_HOT_LOADER__ === 'undefined') {
    return;
  }

  __REACT_HOT_LOADER__.register(Cover, 'Cover', 'src/upload/Cover.jsx');

  __REACT_HOT_LOADER__.register(_default, 'default', 'src/upload/Cover.jsx');
}();

;
